### PR TITLE
Restrict question bank access for users without previewer permissions

### PR DESCRIPTION
**Pull request by nritschel**
This addresses #8850, which was recently brought up by @firasm and others as a high priority concern for them.

The instructorQuestions page, as well as all instructorQuestion pages (including instructorQuestionPreview), are now only accessible for users with course previewer permissions. 

Note that during today's office hours, it was brought up that there will be a number of "dead" links (that lead to 403 error pages) for users with instructor access but no course previewer permissions. Spotting and remove every single one shouldn't be a major concern, but I tried to remove at least the majority (e.g. the questions links in the navigation and in assessmentQuestions). In the process, in noticed that PL is generally fairly inconsistent about when these links are removed, so I added some other checks as well.

### CODE REVIEW

**Comment by nritschel** [comment]
There is some interaction with question sharing, that I thought I had addressed, but there's still a test failing and I'm having some difficulty seeing what exactly is going wrong there (i.e., if this is actually catching a bug or if the test is making some assumptions that no longer hold after the change). 

A second pair of eyes would be very much appreciated. :)

**Comment by nritschel** [comment]
Nevermind, I just accidentally found the issue while looking for something else and was able to fix it. Turns out it was not related to the complicated setup for that particular test at all, but just a generic issue with HTML characters being unintentionally escaped...

File: apps/prairielearn/src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.html.ts:210]:
```suggestion
                    ? html`<a href="${urlPrefix}/question/${question.question_id}/">`
```

Comment (nritschel) [reply apps/prairielearn/src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.html.ts:210]: I tried that (see my earlier commit "Fixed prettier issue with non-closed link"), but then prettier complains that the tag is not closed.

Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.html.ts:210]: Ah, that makes sense. As a rule, `unsafeHtml` should be reserved for exceptional cases. If I can recommend an alternative: this might be a good fit for the `run` function from https://github.com/PrairieLearn/PrairieLearn/pull/10595.

```ts
html`
  ${run(() => {
    const number = question.alternative_group_size === 1
      ? `${question.alternative_group_number}.`
      : html`
          <span class="ml-3">
            ${question.alternative_group_number}.${question.number_in_alternative_group}.
          </span>
        `;
    const title = html`${number} ${question.title}`;

    if (hasCoursePermissionPreview) {
      return html`<a href="${urlPrefix}/question/${question.question_id}/">${title}</a>`;
    }

    return title;
  })}
`;
```

Plus or minus some Prettier formatting, of course.

File: apps/prairielearn/src/pages/instructorQuestions/instructorQuestions.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorQuestions/instructorQuestions.html.ts:47]: This little thing now exists on 4-5 pages with almost no variation (with the exception of the card title and what permission is listed. It might be nice to lift this into a shared `InsufficientCoursePermissionsCard` component in the `components` directory. I won't block on this.

File: apps/prairielearn/src/components/Navbar.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/components/Navbar.html.ts:685]: Nit:

```suggestion
    ${authz_data.has_course_permission_preview
      ? html`
        <li
          class="nav-item ${navPage === 'course_admin' && navSubPage === 'questions'
            ? 'active'
            : ''}"
        >
          <a class="nav-link" href="${urlPrefix}/course_admin/questions">Questions</a>
        </li>
      `
      : ''}
    ${authz_data.has_course_permission_edit
      ? html`
        <li
          class="nav-item ${navPage === 'course_admin' && navSubPage === 'syncs' ? 'active' : ''}"
        >
          <a class="nav-link" href="${urlPrefix}/course_admin/syncs">Sync</a>
        </li>
      `
      : ''}
```

File: apps/prairielearn/src/middlewares/selectAndAuthzInstructorQuestion.sql
Comment (nwalters512) [inline apps/prairielearn/src/middlewares/selectAndAuthzInstructorQuestion.sql:1]: While I think that the conditions here are _technically_ the most correct thing, I'm not sure if we gain much by allowing non-previewers to view publicly shared questions through these routes, especially since we ideally would never render links to them. I'd be comfortable adding an unconditional 403 to `selectAndAuthzInstructorQuestion` if `res.locals.authz_data.has_course_permission_preview === false`. How do others feel?

Candidly, this confused me in local testing too: I just so happened to be viewing a publicly-shared question and was confused for a minute when refreshing didn't 403 after I set my course permissions to "None".

Comment (nritschel) [reply apps/prairielearn/src/middlewares/selectAndAuthzInstructorQuestion.sql:1]: I did have a simpler version before where I just added the pre-defined authzHasCoursePreview check, but then started making this more fine-grained while investigating the breaking test. 

I'm not super attached to this, but considering that the question sharing features might be significantly expanded in the future, I can see some benefit to keeping the finer granularity. For example, I could see some kind of "all public questions database" view becoming a thing in the future (or even just instructors sending direct links to colleagues to show them questions/elements), which then wouldn't require further edits to this page.

Comment (nwalters512) [reply apps/prairielearn/src/middlewares/selectAndAuthzInstructorQuestion.sql:1]: Either of those cases should use the dedicated public viewing routes rather than the course-specific ones. Those don't run the `selectAndAuthzInstructorQuestion` middleware at all:

https://github.com/PrairieLearn/PrairieLearn/blob/771770168c229e1584c9c70acb6b2385b2633c04/apps/prairielearn/src/server.js#L1982-L1989

File: apps/prairielearn/src/pages/instructorQuestions/instructorQuestions.html.ts
Comment (nritschel) [reply apps/prairielearn/src/pages/instructorQuestions/instructorQuestions.html.ts]: Done.

File: apps/prairielearn/src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.html.ts
Comment (nritschel) [reply apps/prairielearn/src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.html.ts]: Nice, this certainly is handy here (I understand that unsafeHtml is not ideal, but without run the alternative would've been even uglier...)

File: apps/prairielearn/src/middlewares/selectAndAuthzInstructorQuestion.sql
Comment (nritschel) [reply apps/prairielearn/src/middlewares/selectAndAuthzInstructorQuestion.sql:1]: Reverted to the more coarse access check

**Review by nwalters512 — COMMENTED** [review]
The permissions checks look good now. Once more comment that's more of a nit than anything.

It would be good to have some tests in place for this, but I won't block on that.

Let me know if/when you're ready for this to be merged.

File: apps/prairielearn/src/components/InsufficientCoursePermissionsCard.ts
Comment (nritschel) [reply apps/prairielearn/src/components/InsufficientCoursePermissionsCard.ts]: What about the edit I made in my commit just now? It gives future pages the option to just use the card, but also doesn't require duplicated code for rendering the header in the current pages.

Comment (nritschel) [reply apps/prairielearn/src/components/InsufficientCoursePermissionsCard.ts]: (Assuming this is acceptable, I am fine to have this merged whenever it looks good to you)

Comment (nritschel) [reply apps/prairielearn/src/components/InsufficientCoursePermissionsCard.ts]: I just saw that you asked for tests above. If there's an obvious place to add a test I'm happy to do that, but after a quick look it seems like PL is generally missing some fine-grained tests for all the different access combinations. Please tell me if I missed something, but the `instructorQuestions.test.ts` tests don't check access, and the `access.test.ts` tests only check student vs. instructor permissions fairly coarsely.

Comment (nwalters512) [reply apps/prairielearn/src/components/InsufficientCoursePermissionsCard.ts]: Splitting up the document and card is good enough for now. I'll still want to change that in the future, but we can save that for another PR.

`instructorQuestions.test.ts` would be a reasonable place to add tests for this. You are correct in that we don't have a single centralized test suite for fine-grained access control.

I'll merge this PR as-is.

**Review by nwalters512 — APPROVED** [review]

**Pull request merged**
